### PR TITLE
v1

### LINKED ISSUE #5: Support just a main command
## Summary

What:

Add support for just a main command that doesn't require subcommands. This should be able to have options as well.

Why:

A user in the discord help channel raised the issue and its a legit use case in my opinion. 

## Acceptance Criteria

Below is a list of tasks that must be completed before this issue can be closed.

- [ ] Write documentation
- [ ] Write unit tests
- [ ] Write integration tests

### LINKED ISSUE #11: Functional improvements to Line
### Discussed in https://github.com/drashland/line/discussions/10

<div type='discussions-op-text'>

<sup>Originally posted by **kinghat** July 22, 2021</sup>
i see there are some additions in the pipeline but maybe we could discuss a few more here?

1. ability to add short names to passed options/flags like `-a` for `--all` or `-l` for `--long`.
2. when passing `--long` it would need a value for the `getOptionValue` call but sometimes options dont need a value and the option being passed is enough.
3. being able to get help on an option on a subcommand or the subcommand itself when the subcommand has logic with or without passing args or options. in that case, passing `-h` would halt the logic and present you the help of args and options for that subcommand.</div>

### PR DESCRIPTION
**Pull request by ebebbington**
Closes #11 
Closes #5 

## Notes

- Introduces v1
- Code is exercised via integration tests (there aren't any unit tests). Should we add unit tests or do the integration tests suffice so far?

## How to create a CLI

- See https://github.com/drashland/line/blob/5-main-command/tests/integration/file_manager/cli.ts
- See https://github.com/drashland/line/blob/5-main-command/tests/integration/file_manager_with_everything/cli.ts
- See https://github.com/drashland/line/blob/5-main-command/tests/integration/file_manager_with_options/cli.ts
- See https://github.com/drashland/line/blob/5-main-command/tests/integration/file_manager_with_too_many_options/cli.ts

There is something that is bothering me. Right now, to describe an argument, you have to provide an `arguments` object. For example:

```diff
class MyCommand extends Line.MainCommand {
  public signature = "line [arg]";

+  public arguments = {
+    "arg": "Some description.",
+  };

  public async handle(): Promise<void> {
    console.log("sup");
  }
}

const cli = new Line.Cli({
  name: "My Cool CLI",
  description: "My cool description",
  version: "v1.0.0",
  command: MyCommand,
});

cli.run();
```

And when you run it, it outputs the following:

```diff
My Cool CLI - My cool description

USAGE

    line [option]
    line [arg: arg]

ARGUMENTS

    arg
+        Some description.

OPTIONS

    -h, --help
        Show this menu.
    -v, --version
        Show this CLI's version.
```

I feel like `arguments` is a weird property name. Maybe `argument_descriptions`? I wouldn't say `descriptions`, because the `options` property also contains descriptions (the value is the description of the option(s)), so it's kind of confusing in that regard. Thoughts though?

## TODO

- [ ] Change arguments description object?
- [ ] Documentation

### CODE REVIEW

File: src/command.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  /**
+   * An array of subcommands that this command can execute.
+   */
+  public subcommands: typeof Line.Subcommand[] = [];
+
+  public cli: Line.Cli;
+
+  public arguments: { [k: string]: string } = {};
+
+  public options: { [k: string]: string } = {};
```
Comment (ebebbington) [inline src/command.ts:21]: Don't suppose we could use maps? just for that tiny bit increase of perf?

File: src/command_line.ts [hunk 2]
```diff
@@ -25,24 +24,17 @@ export class CommandLine {
...
    */
-  public options: { [key: string]: string | undefined } = {};
-
-  /**
-   * The subcommand in this command line. This is the second argument in the
-   * command line. For example, "commit" is the subcommand in the command "git
-   * commit".
-   */
-  public subcommand: string;
+  public options: { [key: string]: string | boolean } = {};
```
Comment (ebebbington) [inline src/command_line.ts:27]: map instead?

File: src/command_line.ts [hunk 3]
```diff
@@ -52,26 +44,65 @@ export class CommandLine {
...
+        value = true;
+      }
+
+      if (value == "true") {
+        value = true;
+      }
+
+      if (value == "false") {
+        value = false;
+      }
```
Comment (ebebbington) [inline src/command_line.ts:100]: what about doing a `value = Boolean(value)`?

File: src/subcommand.ts [hunk 4]
```diff
@@ -1,44 +1,47 @@
...
    */
-  public description = "";
+  public description!: string;
 
   /**
-   * This subcommand's name.
+   * This subcommand's options.
    */
-  public name = "";
+  public options: { [k: string]: string } = {};
```
Comment (ebebbington) [inline src/subcommand.ts:23]: map?

File: src/subcommand.ts [hunk 5]
```diff
@@ -1,44 +1,47 @@
...
+  public options_parsed: string[] = [];
 
   /**
-   * This subcommand's options.
+   * An object that describes the arguments in the signature. If a subcommand
+   * defines this property, then it will be used in the help menu.
    */
-  public options: (typeof SubcommandOption[] | SubcommandOption[]) = [];
+  public arguments: { [k: string]: string } = {};
```
Comment (ebebbington) [inline src/subcommand.ts:30]: map?

**Review by ebebbington — COMMENTED** [review]
lgtm so far

File: src/command.ts (hunk 1 above)
Comment (crookse) [reply src/command.ts:21]: yea it's my plan to optimize this entire project to use maps in all places. key value stores are hella old school now 😏

File: src/command_line.ts (hunk 3 above)
Comment (crookse) [reply src/command_line.ts:100]: for the life of me i couldn't figure out why this code looked so jank. thank you

**Pull request closed without merging**